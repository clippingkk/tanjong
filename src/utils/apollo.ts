import { ApolloClient, ApolloLink, Cache, HttpLink, InMemoryCache } from '@apollo/client'
import { onError } from "@apollo/client/link/error"
import { API_HOST } from '../constants/config'
// import profile from '../utils/profile'
import { offsetLimitPagination } from '@apollo/client/utilities'
import { storage } from './storage'
import toast from 'react-hot-toast/headless'

export type IBaseResponseData = {
  status: number
  msg: string
  data: any
} | {
  code: number,
  error: string // json string
}

export type CKNetworkError = {
  name: string
  response: any
  statusCode: number
  result?: {
    data: null,
    errors: {
      extensions: {
        code: number
        message: string
      }
      message: string
      path: string[]
    }[]
  }
}

let token = getLocalToken()
// let token = localProfile?.token

export function getLocalToken() {
  const tk = storage.getString('me:token')
  if (!tk) {
    return null
  }
  const parsed: string = JSON.parse(tk)
  return parsed.replace(/"/g, '')
}

export function updateLocalToken(nt: string | null) {
  token = nt
}

export async function request<T>(url: string, options: RequestInit = {}): Promise<T> {
  if (token) {
    options.headers = {
      ...(options.headers || {}),
      'Authorization': `Bearer ${token}`
    }
  }
  options.credentials = 'include'
  options.mode = 'cors'

  try {
    const response = await fetch(API_HOST + '/api' + url, options)
    const jsonData = (await response.json()) satisfies IBaseResponseData
    if (response.status >= 400) {
      throw new Error(jsonData.msg ?? jsonData.error)
    }
    if (jsonData.code && jsonData.code >= 400) {
      throw new Error(jsonData.error)
    }
    return jsonData.data as T
  } catch (e) {
    // toast.error('请求挂了... 一会儿再试试')
    throw e
  }
}

const authLink = new ApolloLink((operation, forward) => {
  operation.setContext(({ headers = {} }) => {
    if (!token || token === 'null') {
      return headers
    }

    return {
      headers: {
        ...headers,
        'Authorization': `Bearer ${token}`
      }
    }
  })

  return forward(operation)
})

type GraphQLResponseError = {
  name: string
  response: Response
  result: { code: number, error: string }
  statusCode: number
  message: string
}

const errorLink = onError((errData) => {
  const { graphQLErrors, networkError } = errData
  if (graphQLErrors && graphQLErrors.length > 0) {
    // toast.error(
    //   graphQLErrors[0].message,
    // )
  }
  let ne = networkError as GraphQLResponseError

  if (ne) {
    console.log(`[Network error]: ${ne}`)
    if (ne.statusCode && ne.statusCode === 401) {
      updateLocalToken('')
      // profile.onLogout()
    }
    // toast.error(`${ne.statusCode}: ${ne.name}`)
    // swal({
    //   icon: 'error',
    //   title: `${ne.statusCode}: ${ne.name}`,
    //   text: ne.result?.error
    // })
  }
});

const httpLink = new HttpLink({
  uri: API_HOST + '/api/v2/graphql',
})

export const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          // featuredClippings: offsetLimitPagination(false)
          featuredClippings: {
            keyArgs: false,
            merge(p = [], n) {
              return [...p, ...n]
            }
          },
          books: {
            keyArgs: ['doubanId'],
            merge(p = [], n) {
              return [...p, ...n]
            }
          }
        }
      },
      Book: {
        keyFields: ["doubanId"],
        fields: {
          clippings: {
            merge: simpleDistArrayMerge
          }
        }
      },
      User: {
        fields: {
          recents: {
            merge: simpleDistArrayMerge
          }
        }
      }
    }
  }),
  link: errorLink.concat(authLink.concat(httpLink)),
})

function simpleDistArrayMerge(existings: { __ref: string }[] = [], incoming: { __ref: string }[] = []) {
  return [...existings, ...incoming].reduce((acc, x) => {
    if (acc.findIndex((z: any) => z.__ref === x.__ref) === -1) {
      acc.push(x)
    }
    return acc
  }, [] as any[])
}
