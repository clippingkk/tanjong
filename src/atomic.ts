import { ProfileQuery } from "./schema/generated"
import { atomWithMMKV } from "./utils/jotai"

export const tokenAtom = atomWithMMKV<string | null>('me:token', null)
export const uidAtom = atomWithMMKV<number | null>('me:uid', null)
export const profileAtom = atomWithMMKV<ProfileQuery['me'] | null>('me:profile', null)