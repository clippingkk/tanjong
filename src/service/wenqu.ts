import { WENQU_SIMPLE_TOKEN, WENQU_API_HOST } from "../constants/config"
import { storage } from "../utils/storage"

type WenquErrorResponse = {
  code: number
  error: string
}

export async function wenquRequest<T extends object>(url: string, options: RequestInit = {}): Promise<T> {
  options.headers = {
    ...(options.headers || {}),
    'X-Simple-Check': WENQU_SIMPLE_TOKEN
  }
  options.credentials = 'include'
  options.mode = 'cors'

  try {
    const response: T | WenquErrorResponse = await fetch(WENQU_API_HOST + url, options).then(res => res.json())
    if ('error' in response) {
      throw new Error(response.error)
    }
    return response
  } catch (e) {
    // Sentry.captureException(e)
    return Promise.reject(e)
  }
}

export interface WenquBook {
  id: number
  rating: number
  author: string
  pubdate: string
  totalPages: number
  originTitle: string
  image: string
  doubanId: number
  title: string
  url: string
  press: string
  isbn: string
  tags: string[]
  authorIntro: string
  summary: string
  createdAt: string
  updatedAt: string
}

export interface WenquSearchResponse {
  count: number
  books: WenquBook[]
}

// export class WenquSWRCache implements Cache<WenquBook[]> {
//   private _cache = new Map<string, any>()
//   private _keys = new Set<string>()
//   private prefix = 'wenqu:cache:'
//   private cacheKeysKey = 'wenqu:cache:keys'

//   constructor() {
//     // load keys
//     const lks = storage.getString(this.cacheKeysKey)
//     if (!lks) {
//       return
//     }
//     try {
//       const ks: string[] = JSON.parse(lks)
//       this._keys = new Set(ks)
//     } catch (err) {
//       console.error(err)
//       // do nothing
//     }

//     // TODO: load keys from storage to memory
//   }

//   keys(): IterableIterator<string> {
//     return this._cache.keys()
//   }
//   get(key: Key): State<WenquBook[], any> | undefined {
//     if (!key) {
//       return undefined
//     }
//     return this._cache.get(key as any)
//   }
//   set(key: Key, value: State<WenquBook[], any>): void {
//     this._cache.set(key as any, value)
//   }
//   delete(key: Key): void {
//     this._cache.delete(key as any)
//   }
// }