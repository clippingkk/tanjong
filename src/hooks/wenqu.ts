import { useEffect, useMemo, useRef, useState } from "react"
import useSWR from 'swr'
import { WenquBook, wenquRequest, WenquSearchResponse } from "../service/wenqu"

type bookRequestReturn = {
  books: WenquBook[]
  loading: boolean
}

const cache = new Map<number, WenquBook | null>()

export function useSingleBook(doubanId?: string, skip?: boolean) {
  return useSWR<WenquSearchResponse>( doubanId && `/books/search?dbId=${doubanId}`)
}

// export function useMultipBook(doubanIds: string[], skip?: boolean) {
//   const query = useMemo(() => {
//     if (doubanIds.length === 0) {
//       return null
//     }
//     return doubanIds.join('&dbIds=')
//   }, [doubanIds])

//   return useSWR<WenquSearchResponse>(query && `/books/search?dbIds=${query}`)
// }

export function useBookSearch(query: string, offset: number) {
  const [books, setBooks] = useState<WenquSearchResponse | null>(null)
  useEffect(() => {
    if (query.length < 3) {
      return
    }

    wenquRequest<WenquSearchResponse>(`/books/search?query=${query}&limit=50&offset=${offset}`)
      .then(bs => {
        setBooks(bs)
      })
  }, [query, offset])

  return books
}