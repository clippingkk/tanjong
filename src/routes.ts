import { Clipping } from "./schema/generated"
import { WenquBook } from "./service/wenqu"

export enum RouteKeys {
  AuthV3 = 'AuthV3',
  SignUp = 'Signup',
  AuthAppleBind = 'AuthAppleBind',
  AuthPhoneOTP = 'AuthPhoneOTP',
  AuthQRCode = 'AuthQRCode',
  TabHome = 'TabHome',
  TabProfile = 'TabProfile',
  Book = 'Book',
  Profile = 'Profile',
  Clipping = 'Clipping',
  BookDetail = 'BookDetail',
  ProfileSettings = 'ProfileSettings',
  ProfileDebug = 'ProfileDebug',
  Payment = 'Payment'
}

export type RouteParamList = {
  empty: object
  Book: {
    book: WenquBook
  }
  Clipping: {
    clipping?: Pick<Clipping, 'id' | 'bookID' | 'content' | 'title'>
    clippingID?: number
  }
  AuthAppleBind: {
    idToken: string
  }
  AuthPhoneOTP: {
    phone: string
    idToken: string
  }
  [RouteKeys.AuthQRCode]: object
  [RouteKeys.Payment]: object
  [RouteKeys.AuthV3]: object
  [RouteKeys.SignUp]: object
  [RouteKeys.BookDetail]: {
    book: WenquBook
  }
  [RouteKeys.ProfileSettings]: object
  [RouteKeys.ProfileDebug]: object

  root: object
}

export type TabRouteParamList = {
  [RouteKeys.TabHome]: object
  [RouteKeys.TabProfile]: {}
}
