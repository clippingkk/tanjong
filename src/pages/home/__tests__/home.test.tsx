import 'react-native'
import React from 'react'
import HomePage from '../home.page'
import nock from 'nock'
import { act, render, screen } from '@testing-library/react-native'
import "fast-text-encoding"
import { ApolloProvider } from '@apollo/client'
import { client } from '../../../utils/apollo'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import { RouteKeys } from '../../../routes'
import { NativeBaseProvider, Text } from 'native-base'
import { MockedProvider } from "@apollo/client/testing"
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'jotai'
import { uidAtom } from '../../../atomic'
import { BooksDocument, BooksQuery } from '../../../schema/generated'
import { WenquSearchResponse } from '../../../service/wenqu'

test('home page renders with permission block', async () => {
  const TabStack = createBottomTabNavigator()
  render(
    <ApolloProvider client={client}>
      <NativeBaseProvider initialWindowMetrics={{
        frame: { x: 0, y: 0, width: 0, height: 0 },
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
      }}>
        <NavigationContainer>
          <TabStack.Navigator initialRouteName={RouteKeys.TabHome}>
            <TabStack.Screen
              name={RouteKeys.TabHome}
              component={HomePage}
              options={{
                tabBarLabel: 'Books',
                tabBarIcon: ({ color, size }) => (
                  <Text>ðŸ“š</Text>
                ),
              }}
            />
          </TabStack.Navigator>
        </NavigationContainer>
      </NativeBaseProvider>
    </ApolloProvider>
  )

  expect(screen).toMatchSnapshot()
})

test('home page renders with list', async () => {
  const TabStack = createBottomTabNavigator()
  nock('https://wenqu.annatarhe.cn')
    .get('/books/search?dbId=11')
    .reply(200, {
      count: 1,
      books: [{
        doubanId: 11,
        title: 'hhhhhhhh'
      }, {
        doubanId: 22,
        title: '2222',
      }]
    } as WenquSearchResponse)

  const d = render(
    <MockedProvider mocks={[
      {
        request: {
          query: BooksDocument,
          variables: {
            id: 1,
            pagination: {
              limit: 10,
              offset: 0
            }
          }
        },
        result: {
          data: {
            books: [{
              doubanId: '11',
            }, {
              doubanId: '22'
            }],
            me: {
              id: 1,
              domain: '111',
              recents: [{
                id: 1,
                content: 'lll',
                bookID: '11'
              }, {
                id: 2,
                content: 'rrr',
                bookID: '22',
              }],
              recent3mReadings: ['11']
            }
          } as BooksQuery
        }
      }
    ]}>
      <NativeBaseProvider initialWindowMetrics={{
        frame: { x: 0, y: 0, width: 0, height: 0 },
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
      }}>
        <Provider initialValues={[[uidAtom, 1]]}>
          <NavigationContainer>
            <TabStack.Navigator initialRouteName={RouteKeys.TabHome}>
              <TabStack.Screen
                name={RouteKeys.TabHome}
                component={HomePage}
                options={{
                  tabBarLabel: 'Books',
                  tabBarIcon: ({ color, size }) => (
                    <Text>ðŸ“š</Text>
                  ),
                }}
              />
            </TabStack.Navigator>
          </NavigationContainer>
        </Provider>
      </NativeBaseProvider>
    </MockedProvider>
  )
  // make sure every thing rendered
  await act(() => {})
  expect(screen).toMatchSnapshot()
})
