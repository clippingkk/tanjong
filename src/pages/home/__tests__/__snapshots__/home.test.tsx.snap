// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page renders with list 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenNavigationContainer
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RNSScreen
        activityState={2}
        collapsable={false}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        onGestureCancel={[Function]}
        sheetAllowedDetents="large"
        sheetCornerRadius={-1}
        sheetExpandsWhenScrolledToEdge={true}
        sheetGrabberVisible={false}
        sheetLargestUndimmedDetent="all"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            [
              {
                "backgroundColor": "rgb(242, 242, 242)",
                "flex": 1,
              },
              [
                {
                  "flex": 1,
                  "flexDirection": "column-reverse",
                },
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <View
              data-style=""
              dataSet={
                {
                  "style": "",
                }
              }
              justifyItems="center"
              style={
                [
                  undefined,
                  {},
                  {},
                  undefined,
                  {
                    "backgroundColor": "#f1f5f9",
                    "flex": 1,
                    "justifyContent": "center",
                    "minHeight": 1334,
                  },
                  undefined,
                ]
              }
              sx={
                {
                  "_dark": {
                    "backgroundColor": "$blueGray900",
                  },
                }
              }
            >
              <View
                data-style=""
                dataSet={
                  {
                    "style": "",
                  }
                }
                style={
                  [
                    undefined,
                    {},
                    {},
                    undefined,
                    {
                      "alignItems": "center",
                      "flex": 1,
                      "justifyContent": "center",
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  data-style=""
                  dataSet={
                    {
                      "style": "",
                    }
                  }
                  size="md"
                  style={
                    [
                      undefined,
                      {},
                      {
                        "color": "#525252",
                        "fontFamily": undefined,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                      },
                      undefined,
                      {
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 48,
                        "textAlign": "center",
                        "width": "100%",
                      },
                      undefined,
                    ]
                  }
                >
                  â›”
                </Text>
                <Text
                  accessibilityRole="link"
                  href="/AuthV3"
                  onPress={[Function]}
                  style={
                    {
                      "textAlign": "center",
                      "width": "100%",
                    }
                  }
                >
                  <Text
                    data-style=""
                    dataSet={
                      {
                        "style": "",
                      }
                    }
                    size="md"
                    style={
                      [
                        undefined,
                        {},
                        {
                          "color": "#525252",
                          "fontFamily": undefined,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                        },
                        undefined,
                        {
                          "fontSize": 16,
                          "lineHeight": 22,
                        },
                        {
                          "fontSize": 16,
                          "lineHeight": 22,
                        },
                        {
                          "color": "#0f172a",
                        },
                        undefined,
                      ]
                    }
                    sx={
                      {
                        "_dark": {
                          "color": "$blueGray100",
                        },
                      }
                    }
                  >
                    app.auth.loginFullTip
                  </Text>
                </Text>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "zIndex": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgb(255, 255, 255)",
                    "borderBottomColor": "rgb(216, 216, 216)",
                    "flex": 1,
                    "shadowColor": "rgb(216, 216, 216)",
                    "shadowOffset": {
                      "height": 0.5,
                      "width": 0,
                    },
                    "shadowOpacity": 0.85,
                    "shadowRadius": 0,
                  }
                }
              />
            </View>
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "height": 44,
                  "maxHeight": undefined,
                  "minHeight": undefined,
                  "opacity": undefined,
                  "transform": undefined,
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  {
                    "height": 0,
                  }
                }
              />
              <View
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "stretch",
                    "flex": 1,
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "alignItems": "flex-start",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "justifyContent": "center",
                      "marginStart": 0,
                    }
                  }
                />
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "justifyContent": "center",
                      "marginHorizontal": 16,
                      "maxWidth": 718,
                    }
                  }
                >
                  <Text
                    accessibilityRole="header"
                    aria-level="1"
                    collapsable={false}
                    numberOfLines={1}
                    style={
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 17,
                        "fontWeight": "600",
                      }
                    }
                  >
                    TabHome
                  </Text>
                </View>
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "alignItems": "flex-end",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "justifyContent": "center",
                      "marginEnd": 0,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenNavigationContainer>
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderTopColor": "rgb(216, 216, 216)",
          "borderTopWidth": 0.5,
          "bottom": 0,
          "elevation": 8,
          "height": 49,
          "left": 0,
          "paddingBottom": 0,
          "paddingHorizontal": 0,
          "position": null,
          "right": 0,
          "transform": [
            {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <View
        accessibilityRole="tablist"
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityLabel="Books, tab, 1 of 1"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": true,
            }
          }
          accessibilityStates={
            [
              "selected",
            ]
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "flex": 1,
              },
              {
                "backgroundColor": "transparent",
              },
              {
                "flexDirection": "column",
                "justifyContent": "flex-end",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "minWidth": 25,
                    "position": "absolute",
                    "width": "100%",
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <Text
                dataSet={{}}
                style={
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                ðŸ“š
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "minWidth": 25,
                    "position": "absolute",
                    "width": "100%",
                  },
                  {
                    "opacity": 0,
                  },
                ]
              }
            >
              <Text
                dataSet={{}}
                style={
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                ðŸ“š
              </Text>
            </View>
          </View>
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "textAlign": "center",
                },
                {
                  "color": "rgb(0, 122, 255)",
                },
                {
                  "fontSize": 10,
                },
                undefined,
              ]
            }
          >
            Books
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`home page renders with permission block 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenNavigationContainer
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RNSScreen
        activityState={2}
        collapsable={false}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        onGestureCancel={[Function]}
        sheetAllowedDetents="large"
        sheetCornerRadius={-1}
        sheetExpandsWhenScrolledToEdge={true}
        sheetGrabberVisible={false}
        sheetLargestUndimmedDetent="all"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            [
              {
                "backgroundColor": "rgb(242, 242, 242)",
                "flex": 1,
              },
              [
                {
                  "flex": 1,
                  "flexDirection": "column-reverse",
                },
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              {
                "flex": 1,
              }
            }
          >
            <View
              data-style=""
              dataSet={
                {
                  "style": "",
                }
              }
              justifyItems="center"
              style={
                [
                  undefined,
                  {},
                  {},
                  {},
                  {
                    "backgroundColor": "#f1f5f9",
                    "flex": 1,
                    "justifyContent": "center",
                    "minHeight": 1334,
                  },
                  undefined,
                ]
              }
              sx={
                {
                  "_dark": {
                    "backgroundColor": "$blueGray900",
                  },
                }
              }
            >
              <View
                data-style=""
                dataSet={
                  {
                    "style": "",
                  }
                }
                style={
                  [
                    undefined,
                    {},
                    {},
                    {},
                    {
                      "alignItems": "center",
                      "flex": 1,
                      "justifyContent": "center",
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  data-style=""
                  dataSet={
                    {
                      "style": "",
                    }
                  }
                  size="md"
                  style={
                    [
                      undefined,
                      {},
                      {
                        "color": "#525252",
                        "fontFamily": undefined,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                      },
                      {},
                      {
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 48,
                        "textAlign": "center",
                        "width": "100%",
                      },
                      undefined,
                    ]
                  }
                >
                  â›”
                </Text>
                <Text
                  accessibilityRole="link"
                  href="/AuthV3"
                  onPress={[Function]}
                  style={
                    {
                      "textAlign": "center",
                      "width": "100%",
                    }
                  }
                >
                  <Text
                    data-style=""
                    dataSet={
                      {
                        "style": "",
                      }
                    }
                    size="md"
                    style={
                      [
                        undefined,
                        {},
                        {
                          "color": "#525252",
                          "fontFamily": undefined,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                        },
                        {},
                        {
                          "fontSize": 16,
                          "lineHeight": 22,
                        },
                        {
                          "fontSize": 16,
                          "lineHeight": 22,
                        },
                        {
                          "color": "#0f172a",
                        },
                        undefined,
                      ]
                    }
                    sx={
                      {
                        "_dark": {
                          "color": "$blueGray100",
                        },
                      }
                    }
                  >
                    app.auth.loginFullTip
                  </Text>
                </Text>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "zIndex": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgb(255, 255, 255)",
                    "borderBottomColor": "rgb(216, 216, 216)",
                    "flex": 1,
                    "shadowColor": "rgb(216, 216, 216)",
                    "shadowOffset": {
                      "height": 0.5,
                      "width": 0,
                    },
                    "shadowOpacity": 0.85,
                    "shadowRadius": 0,
                  }
                }
              />
            </View>
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "height": 44,
                  "maxHeight": undefined,
                  "minHeight": undefined,
                  "opacity": undefined,
                  "transform": undefined,
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  {
                    "height": 0,
                  }
                }
              />
              <View
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "stretch",
                    "flex": 1,
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "alignItems": "flex-start",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "justifyContent": "center",
                      "marginStart": 0,
                    }
                  }
                />
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "justifyContent": "center",
                      "marginHorizontal": 16,
                      "maxWidth": 718,
                    }
                  }
                >
                  <Text
                    accessibilityRole="header"
                    aria-level="1"
                    collapsable={false}
                    numberOfLines={1}
                    style={
                      {
                        "color": "rgb(28, 28, 30)",
                        "fontSize": 17,
                        "fontWeight": "600",
                      }
                    }
                  >
                    TabHome
                  </Text>
                </View>
                <View
                  collapsable={false}
                  pointerEvents="box-none"
                  style={
                    {
                      "alignItems": "flex-end",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "justifyContent": "center",
                      "marginEnd": 0,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenNavigationContainer>
    <View
      collapsable={false}
      onLayout={[Function]}
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "rgb(255, 255, 255)",
          "borderTopColor": "rgb(216, 216, 216)",
          "borderTopWidth": 0.5,
          "bottom": 0,
          "elevation": 8,
          "height": 49,
          "left": 0,
          "paddingBottom": 0,
          "paddingHorizontal": 0,
          "position": null,
          "right": 0,
          "transform": [
            {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <View
        accessibilityRole="tablist"
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityLabel="Books, tab, 1 of 1"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": true,
            }
          }
          accessibilityStates={
            [
              "selected",
            ]
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "flex": 1,
              },
              {
                "backgroundColor": "transparent",
              },
              {
                "flexDirection": "column",
                "justifyContent": "flex-end",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "minWidth": 25,
                    "position": "absolute",
                    "width": "100%",
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <Text
                dataSet={{}}
                style={
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                ðŸ“š
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "minWidth": 25,
                    "position": "absolute",
                    "width": "100%",
                  },
                  {
                    "opacity": 0,
                  },
                ]
              }
            >
              <Text
                dataSet={{}}
                style={
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  }
                }
              >
                ðŸ“š
              </Text>
            </View>
          </View>
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "textAlign": "center",
                },
                {
                  "color": "rgb(0, 122, 255)",
                },
                {
                  "fontSize": 10,
                },
                undefined,
              ]
            }
          >
            Books
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
